### ISSUE TITLE
🐛 `lint/complexity/noForEach` false-positive for non `Array` methods

### ISSUE THREAD

**Issue by danielsarsi** [biomejs/biome#1205]
### Environment information

```block
CLI:
  Version:                      1.4.1
  Color support:                true

Platform:
  CPU Architecture:             x86_64
  OS:                           linux

Environment:
  BIOME_LOG_DIR:                unset
  NO_COLOR:                     unset
  TERM:                         "xterm-256color"
  JS_RUNTIME_VERSION:           "v20.8.0"
  JS_RUNTIME_NAME:              "node"
  NODE_PACKAGE_MANAGER:         "yarn/4.0.2"

Biome Configuration:
  Status:                       Loaded successfully
  Formatter disabled:           false
  Linter disabled:              false
  Organize imports disabled:    true
  VCS disabled:                 true

Workspace:
  Open Documents:               0
```

### What happened?

1. Create an object with a method called `forEach`;
2. Use this method;
3. Biome suggests the use of `for...of`, which it's not possible.

```ts
const NotArray = {
  forEach: (str: string, callback: (char: string, index: number) => void) => {
    for (let i = 0; i < str.length; i++) {
      callback(str[i], i);
    }
  },
};

const message = 'This is a message.';

NotArray.forEach(message, (char, index) => {
  console.log(char, index);
});
// ^ Prefer for...of instead of Array.forEach
```

[Playground link.](https://biomejs.dev/playground/?code=YwBvAG4AcwB0ACAATgBvAHQAQQByAHIAYQB5ACAAPQAgAHsACgAgACAAZgBvAHIARQBhAGMAaAA6ACAAKABzAHQAcgA6ACAAcwB0AHIAaQBuAGcALAAgAGMAYQBsAGwAYgBhAGMAawA6ACAAKABjAGgAYQByADoAIABzAHQAcgBpAG4AZwAsACAAaQBuAGQAZQB4ADoAIABuAHUAbQBiAGUAcgApACAAPQA%2BACAAdgBvAGkAZAApACAAPQA%2BACAAewAKACAAIAAgACAAZgBvAHIAIAAoAGwAZQB0ACAAaQAgAD0AIAAwADsAIABpACAAPAAgAHMAdAByAC4AbABlAG4AZwB0AGgAOwAgAGkAKwArACkAIAB7AAoAIAAgACAAIAAgACAAYwBhAGwAbABiAGEAYwBrACgAcwB0AHIAWwBpAF0ALAAgAGkAKQA7AAoAIAAgACAAIAB9AAoAIAAgAH0ALAAKAH0AOwAKAAoAYwBvAG4AcwB0ACAAbQBlAHMAcwBhAGcAZQAgAD0AIAAnAFQAaABpAHMAIABpAHMAIABhACAAbQBlAHMAcwBhAGcAZQAuACcAOwAKAAoATgBvAHQAQQByAHIAYQB5AC4AZgBvAHIARQBhAGMAaAAoAG0AZQBzAHMAYQBnAGUALAAgACgAYwBoAGEAcgAsACAAaQBuAGQAZQB4ACkAIAA9AD4AIAB7AAoAIAAgAGMAbwBuAHMAbwBsAGUALgBsAG8AZwAoAGMAaABhAHIALAAgAGkAbgBkAGUAeAApADsACgB9ACkAOwA%3D)

### Expected result

The rule valid only for `Array.forEach`.

### Code of Conduct

- [X] I agree to follow Biome's Code of Conduct

**Comment by Conaclos** [comment]
I am not sure that we should fix this issue.
The reasoning behind the rule was that if a `forEach` method is available, then the object should be iterable.

If we want to handle this kind of false positive, a type system is required.

**Comment by ematipico** [comment]
> I am not sure that we should fix this issue.
> The reasoning behind the rule was that if a `forEach` method is available, then the object should be iterable.
> 
> If we want to handle this kind of false positive, a type system is required.

Do we document this limitation?

**Comment by Conaclos** [comment]
We didn't in the first place, because the equivalent ESLint rule has the same issue behavior. However, I agree that this could make sense of adding a caveat note in the documentation.
